### ISSUE TITLE
BufReader Stdin or StdinLock

### ISSUE THREAD

**Issue by pickfire** [rust-lang/rust-clippy#6755]
### What it does

Suggest using `StdinLock` rather than `BufReader::new(io::stdin())` since it is already buffered. We might also want to take care of `let stdin = io::stdin(); BufReader::new(stdin)` and also `BufReader::new(stdinlock)` (there could be people doing this).

### Categories (optional)

- Kind: perf

Locking stdin also provide benefit for not having to re-lock it on every read.

> std::io::stdin() already is buffered. Re-buffering it only increase the memcpy calls.

Source https://users.rust-lang.org/t/correct-way-for-taking-large-input-from-keyboard/55304/11
Related to https://github.com/rust-lang/rust-clippy/issues/1805

For example:
- Recommend locking of stdin to reduce re-locking
- Suggest not re-buffering stdin lock which already implements `BufRead`

### Drawbacks

None.

### Example

```rust
let reader = BufReader::new(io::stdin());  // this could a separate variable or StdinLock
```

Could be written as:

```rust
let stdin = io::stdin();
let reader = BufReader::new(stdin.lock());
```

**Comment by camsteffen** [comment]
Shouldn't `BufReader::new(io::stdin())` be replaced by `io::stdin().lock()`? `StdinLock` implements `BufRead`.

**Comment by pickfire** [comment]
Yes, I wrote it wrongly. I made the same mistake which the lint should solve.

**Comment by camsteffen** [comment]
I think this should also include `BufWriter::new(std::io::stdout())`.

**Comment by pickfire** [comment]
@camsteffen I don't think so as `Stdout` does not implements `BufWrite`.

**Comment by camsteffen** [comment]
Ah you're right.

**Comment by CBenoit** [comment]
Hi @arya-k! Do you still plan to work on that?

**Comment by arya-k** [comment]
Hi! I don’t think I have the bandwidth unfortunately- feel free to take over. On Nov 13, 2023, at 21:04, Benoît Cortier ***@***.***> wrote:﻿
Hi @arya-k! Do you still plan to work on that?

**Comment by CBenoit** [comment]
Sounds good, I’ll take to claim it

@rustbot claim

**Comment by yawara** [comment]
Hi @CBenoit , are you still working on this issue?

**Comment by CBenoit** [comment]
> Hi, are you still working on this issue? 

I completely forgot about it. Feel free to claim!

**Comment by yawara** [comment]
@CBenoit Thank you! I claim it.

@rustbot claim
